oapen-engine/src/data/oapen_db.py

```diff
@@ -1,54 +1,136 @@
-from logging import Logger
+from typing import List
 
-from model.oapen_types import OapenSuggestion
+import psycopg2
+from data.connection import connection
+from model.oapen_types import OapenNgram
 
 
-def table_exists(connection, table):
+def table_exists(table):
     cursor = connection.cursor
-    cursor.execute(
-        """
-        SELECT EXISTS (
-            SELECT * FROM oapen_suggestions.tables WHERE table_name=%s
-        )
-        """,
-        (table),
-    )
 
-    res = cursor.fetchone()[0]
-    cursor.close()
-    return res
+    query = """
+            SELECT EXISTS (
+                SELECT * FROM oapen_suggestions.tables WHERE table_name=%s
+            )
+            """
+
+    try:
+        cursor.execute(query, (table))
+        res = cursor.fetchone()[0]
+        return res
+    except (Exception, psycopg2.Error) as error:
+        print(error)
+    finally:
+        cursor.close()
 
 
-def add_suggestion(connection, suggestion: OapenSuggestion) -> None:
+def add_single_suggestion(suggestion) -> None:
     cursor = connection.cursor()
 
     query = """
-            INSERT INTO oapen_suggestions.suggestions VALUES (%s, %s, %s)
-        """
+            INSERT INTO oapen_suggestions.suggestions (handle, name, suggestions)
+            VALUES (%s, %s, %s::oapen_suggestions.suggestion[])
+            ON CONFLICT (handle)
+            DO
+                UPDATE SET suggestions = excluded.suggestions
+            """
 
     try:
-        cursor.execute(query, suggestion)
-    except Exception as ex:
-        Logger.exception(ex)
+        cursor.execute(query, (suggestion[0], suggestion[1], suggestion[2]))
+    except (Exception, psycopg2.Error) as error:
+        print(error)
     finally:
         cursor.close()
 
 
-def add_many_suggestions(connection, suggestions) -> None:
+def add_many_suggestions(suggestions) -> None:
```

Comment:
This is nitpicky in advance, but I just wanted to understand why add many doesn't iteratively just call add one, since it seems like we could condense the code a bit to make it more readable. However, I know sometimes this actually is faster than doing it that way, so feel free to disregard if so.

Reply:
Executing a single query is many times faster than psycopg2's `execute_all` or `execute_batch`